pandas/core/indexes/datetimelike.py

```diff
@@ -583,6 +583,22 @@ def delete(self, loc):
         arr = type(self._data)._simple_new(new_i8s, dtype=self.dtype, freq=freq)
         return type(self)._simple_new(arr, name=self.name)
 
+    # --------------------------------------------------------------------
+    # Join/Set Methods
+
+    def _wrap_joined_index(self, joined: np.ndarray, other):
+        assert other.dtype == self.dtype, (other.dtype, self.dtype)
+        name = get_op_result_name(self, other)
+
+        if is_period_dtype(self.dtype):
+            freq = self.freq
+        else:
+            self = cast(DatetimeTimedeltaMixin, self)
+            freq = self.freq if self._can_fast_union(other) else None
```

Comment:
to avoid this pattern in the base class, could _can_fast_union be overridden in PeriodIndex to always return True.

Reply:
on second thoughts perhaps put in DatetimeIndexOpsMixin and then it's overridden in DatetimeTimedeltaMixin and the cast is not necessary.

Reply:
i considered this, but im planning to implement an actual _can_fast_union for PeriodIndex before long